torch/_inductor/codecache.py

```diff
@@ -249,35 +253,151 @@ def code_hash(code):
+    hashed_fx_args = base64.b32encode(hashlib.sha256(serialized_fx_args).digest())[:51]
+    return "f" + hashed_fx_args.decode("utf-8").lower()
+
+
+def get_path(basename: str, extension: str):
     subdir = os.path.join(cache_dir(), basename[1:3])
-    path = os.path.join(subdir, f"{basename}.{ext}")
-    return extra + basename, subdir, path
+    path = os.path.join(subdir, f"{basename}.{extension}")
+    return basename, subdir, path
 
 
-def write(source_code: Union[str, bytes], ext: str, extra="", binary=False):
-    basename, subdir, path = get_code_path(source_code, ext, extra, binary)
+def write(content: Union[str, bytes], key: str, extension: str):
+    basename, subdir, path = get_path(key, extension)
     if not os.path.exists(subdir):
         os.makedirs(subdir, exist_ok=True)
     if not os.path.exists(path):
-        write_atomic(path, source_code, binary)
+        write_atomic(path, content)
     return basename, path
 
 
-def write_atomic(path: str, source_code: Union[str, bytes], binary=False):
+def write_atomic(path: str, content: Union[str, bytes]):
     # Write into temporary file first to avoid conflicts between threads
     # Avoid using a named temporary file, as those have restricted permissions
+    assert isinstance(
+        content, (str, bytes)
+    ), "Only strings and byte arrays can be saved in the cache"
     path = pathlib.Path(path)
     tmp_path = path.parent / f".{os.getpid()}.{threading.get_ident()}.tmp"
-    with tmp_path.open("wb" if binary else "w") as f:
-        f.write(source_code)
-
+    write_mode = "w" if isinstance(content, str) else "wb"
+    with tmp_path.open(write_mode) as f:
+        f.write(content)
     tmp_path.rename(path)
 
 
+class FxGraphCache:
+    cache = dict()
+    clear = staticmethod(cache.clear)
+
+    @classmethod
+    def load(
+        cls, compile_fx_fn: Callable, fx_args: List[Any], fx_kwargs: Dict[str, Any]
+    ):
+        fx_args_for_hashing = copy(fx_args)
+        fx_args_for_hashing.extend(list(fx_kwargs.values()))
+        # Hash also on torch version and current triton config
+        fx_args_for_hashing.append(torch.__version__)
+        fx_args_for_hashing.append(config.save_config())
+        key = compiled_fx_graph_hash(fx_args_for_hashing)
+        if key not in cls.cache:
+            from filelock import FileLock
+
+            lock_dir = get_lock_dir()
+            lock = FileLock(os.path.join(lock_dir, key + ".lock"), timeout=LOCK_TIMEOUT)
+            with lock:
+                _, _, cg_path = get_path(key, "cg")
+                if not os.path.exists(cg_path):
+                    compiled_graph: CompiledFxGraph = compile_fx_fn(
+                        *fx_args, **fx_kwargs
+                    )
+
+                    disk_compiled_graph = copy(compiled_graph)
+                    # Important as compiled models are not pickeable
+                    disk_compiled_graph.compiled_artifact = None
```

Comment:
soon hopefully this should work https://github.com/pytorch/pytorch/pull/101997 - still debugging a graph break there but will note to revisit this in case your PR gets merged first

Reply:
I see it hasn't been landed yet, so for now I'll keep it there with a TODO to check for #101651